### PR TITLE
Add ability to use custom TLD list

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to use custom TLD list in functions `firstSignificantSubdomainCustom`, `cutToFirstSignificantSubdomainCustom`.

Use cases:
- more frequent syncing (with, i.e. https://publicsuffix.org/)
- custom TLD list

Benchmarks

<details>

```sql
SELECT cutToFirstSignificantSubdomainCustom(URL, 'public_suffix_list')
FROM datasets.hits
SETTINGS max_threads = 1
FORMAT Null
SETTINGS max_threads = 1

0 rows in set. Elapsed: 1.024 sec. Processed 8.87 million rows, 762.68 MB (8.67 million rows/s., 745.11 MB/s.) 

```

```sql
SELECT cutToFirstSignificantSubdomain(URL)
FROM datasets.hits
SETTINGS max_threads = 1
FORMAT Null
SETTINGS max_threads = 1

0 rows in set. Elapsed: 0.978 sec. Processed 8.87 million rows, 762.68 MB (9.08 million rows/s., 779.96 MB/s.) 
```

```
2020.12.04 07:33:34.663548 [ 483105 ] {} <Trace> TLDListsHolder: public_suffix_list loading from public_suffix_list.dat                                                                                                                                        
2020.12.04 07:33:34.666505 [ 483105 ] {} <Information> TLDListsHolder: public_suffix_list was added (9084 hosts)   
```
</details>

### CODE REVIEW

**Comment by azat** [comment]
>Performance — 2 errors, 5 faster, 7 slower, 24 unstable

```
DB::Exception: Unknown function firstSignificantSubdomainCustom. Maybe you meant: ['firstSignificantSubdomain','cutToFirstSignificantSubdomain']: While processing SELECT count() FROM hits_100m_single WHERE NOT ignore(firstSignificantSubdomainCustom(URL, 'public_suffix_list')). Stack trace:
```

Since there is no such function in upstream/master version, small benchmark attached to the PR header by the way.

File: src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#pragma once
```
Comment (akuzm) [inline src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h:1]: It's an implementation detail of the lookup function, why does it need its own header?

File: tests/performance/custom_tld.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<test>
+    <preconditions>
+        <table_exists>hits_100m_single</table_exists>
+    </preconditions>
+
+    <query>SELECT count() FROM hits_100m_single WHERE NOT ignore(firstSignificantSubdomain(URL))</query>
```
Comment (akuzm) [inline tests/performance/custom_tld.xml:6]: Maybe we can put it into `first_significant_subdomain` or `url_hits`? Probably the latter, and the former can be deleted. For the perf tests, having one relatively big test with lots of related functions is often preferrable, because it's more convenient to compare the results, and the aggregate metric over the entire test is meaningful to some extent, e.g. 'how much faster is working with URLs'. Also we only run a subset of queries per test in PRs (no more than 15 queries IIRC), so adding more queries to a big test doesn't make it slower for PRs at least, only for master where we run everything.

File: src/Common/TLDListsHolder.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+private:
+    Container tld_container;
+    std::unique_ptr<Arena> pool;
+};
+
+class TLDListsHolder
+{
+public:
+    /// FIXME: use dense_hash_map and better hash function
```
Comment (akuzm) [inline src/Common/TLDListsHolder.h:39]: This is for storing the list of custom TLD lists, it's small and we search it once per block, so probably any map will do?

File: programs/server/config.xml [hunk 4]
```diff
@@ -714,6 +714,15 @@
     <!-- <path_to_regions_names_files>/opt/geo/</path_to_regions_names_files> -->
 
 
+    <!-- Custom TLD lists.
+         Format: <name>/path/to/file</name>
+    -->
+    <top_level_domains_lists>
+        <!--
+        <public_suffix_list>/path/to/public_suffix_list.dat</public_suffix_list>
```
Comment (akuzm) [inline programs/server/config.xml:722]: Updating a list file requires server restart, right?

Removing a list from configuration looks problematic -- the reference to it from a simultaneously running function may become dangling.

File: src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h (hunk 1 above)
Comment (azat) [reply src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h:1]: Just to make it similar to `src/Functions/URL/ExtractFirstSignificantSubdomainCustomLookup.h`

File: tests/performance/custom_tld.xml (hunk 2 above)
Comment (azat) [reply tests/performance/custom_tld.xml:6]: >Maybe we can put it into first_significant_subdomain or url_hits? Probably the latter, and the former can be deleted. 

Sure. So are you suggesting to move custom_tld.xml and first_significant_subdomain.xml to the url_hits.xml and remove old files?

>For the perf tests, having one relatively big test with lots of related functions is often preferrable, because it's more convenient to compare the results, and the aggregate metric over the entire test is meaningful to some extent, e.g. 'how much faster is working with URLs'

I guess it is better for seeing the picture "in general"

> Also we only run a subset of queries per test in PRs (no more than 15 queries IIRC), so adding more queries to a big test doesn't make it slower for PRs at least, only for master where we run everything.

But this means that new functions will not be even checked in performance tests? Then maybe it will be better to merge them in a separate PR? (since currently this one contains real numbers for those functions, while after it may even skip them, since url_hits.xml already contains 15+ queries)

File: src/Common/TLDListsHolder.h (hunk 3 above)
Comment (azat) [reply src/Common/TLDListsHolder.h:39]: Yep, sure, that's why it is just a comment (I leave this comment only because unordered_set is way more less optimal then dense_hash_map and I couldn't resist)

File: programs/server/config.xml (hunk 4 above)
Comment (azat) [reply programs/server/config.xml:722]: >Updating a list file requires server restart, right?

Correct

>Removing a list from configuration looks problematic -- the reference to it from a simultaneously running function may become dangling.

Actually this can be supported by storing shared_ptr, but this can be done later I guess.

File: src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h (hunk 1 above)
Comment (akuzm) [reply src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h:1]: That one doesn't look like it deserves a header either :)

File: tests/performance/custom_tld.xml (hunk 2 above)
Comment (akuzm) [reply tests/performance/custom_tld.xml:6]: > Sure. So are you suggesting to move custom_tld.xml and first_significant_subdomain.xml to the url_hits.xml and remove old files?

Yep.

> But this means that new functions will not be even checked in performance tests?

This logic is disabled when the PR changes the test files, and we run all the queries.

File: src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h (hunk 1 above)
Comment (azat) [reply src/Functions/URL/ExtractFirstSignificantSubdomainDefaultLookup.h:1]: Agree.

TL;DR; at first it was included into two files, however after few rebases it goes away.

File: programs/server/config.xml (hunk 4 above)
Comment (akuzm) [reply programs/server/config.xml:722]: Somehow I thought we already can remove the list w/o restart, but apparently not, so no problem with dangling references.

File: tests/performance/custom_tld.xml (hunk 2 above)
Comment (azat) [reply tests/performance/custom_tld.xml:6]: >This logic is disabled when the PR changes the test files, and we run all the queries.

Ok, then let's merge those

**Comment by azat** [comment]
>Performance — 2 errors, 8 faster, 7 slower, 37 unstable

Now perf tests looks ok, if you want me to resolve the conflicts please let me know

**Comment by azat** [comment]
>Now perf tests looks ok

No it's not, missing TLD list in perf tests:

```
2020.12.08 17:13:50.007167 [ 258 ] {url_hits.query22.prewarm0} <Error> TCPHandler: Code: 562, e.displayText() = DB::Exception: TLD list public_suffix_list does not exist: while executing 'FUNCTION cutToFirstSignificantSubdomainCustom(URL :: 0, 'public_suffix_list' :: 1) -> cutToFirstSignificantSubdomainCustom(URL, 'public_suffix_list') String : 2', Stack trace:
```

Will take a look

File: tests/config/install.sh [hunk 5]
```diff
@@ -41,6 +42,12 @@ ln -sf $SRC_PATH/strings_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/decimals_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/executable_dictionary.xml $DEST_SERVER_PATH/
 
+# For CI when it is runned via init service
+mkdir -p /var/lib/clickhouse/cores
```
Comment (akuzm) [inline tests/config/install.sh:46]: By the way, what is this `cores` dir and why do we need it? Maybe we can handle the paths the same way we do for `server.key` etc? To have less default dirs with configs.

**Comment by akuzm** [comment]
> No it's not, missing TLD list in perf tests:

It's here:
https://github.com/ClickHouse/ClickHouse/blob/master/docker/test/performance-comparison/compare.sh#L36
It doesn't install the test config, maybe it should do it.

File: tests/config/install.sh (hunk 5 above)
Comment (azat) [reply tests/config/install.sh:46]: >By the way, what is this cores dir and why do we need it?

`/var/lib/clickhouse/cores` is the cwd of the `clickhouse-serv` process under sysv-init (that is used for start `clickhouse-server` in CI except for fasttest)

> Maybe we can handle the paths the same way we do for server.key etc? To have less default dirs with configs.

But `server.key` is installed directly into the `/etc/clickhouse-server`, while I want to run those tests locally, and by not using absolute path to the list in `top_level_domains_lists.xml` it works now

**Comment by azat** [comment]
>It doesn't install the test config, maybe it should do it.

But is sources available in that container?
If not, is there internet in that container?

File: tests/config/install.sh (hunk 5 above)
Comment (akuzm) [reply tests/config/install.sh:46]: So it's dependent on the `init` system? This looks somewhat fragile. Maybe we can default to something in the data directory? Like `user_file` and `format_schemas`, we could use `/var/lib/clickhouse/top_level_domains` or something. Then, to run the tests locally, we can add an override to `programs/server/config.d/path.xml`.

**Comment by azat** [comment]
Ok, I see it `COPY * /`, so I guess I can put a symlink to this file into the `docker/test/performance-comparison`

https://github.com/ClickHouse/ClickHouse/blob/f6f336ac7839ed52a147eb0bcae8b88f0548082a/docker/test/performance-comparison/Dockerfile#L40

File: tests/config/install.sh (hunk 5 above)
Comment (azat) [reply tests/config/install.sh:46]: >So it's dependent on the init system? This looks somewhat fragile.

Indeed.

>Maybe we can default to something in the data directory? Like user_file and format_schemas, we could use /var/lib/clickhouse/top_level_domains or something. Then, to run the tests locally, we can add an override to programs/server/config.d/path.xml.

Looks a little bit overkill for this task, but maybe you right.

Comment (azat) [reply tests/config/install.sh:46]: @akuzm what do you think about just using user_files for this?

Comment (akuzm) [reply tests/config/install.sh:46]: > @akuzm what do you think about just using user_files for this?

This means they can be overwritten by users through SQL queries, not sure if this is OK or not.

Comment (azat) [reply tests/config/install.sh:46]: >This means they can be overwritten by users through SQL queries, not sure if this is OK or not.

You mean something like `INTO OUTFILE '/path/to/user_files/list.dat'`? Yeah, this should not be allowed, ok, thanks

Comment (azat) [reply tests/config/install.sh]: But fasttest will still be different, since it uses another data path

File: tests/config/install.sh [hunk 6]
```diff
@@ -41,6 +42,9 @@ ln -sf $SRC_PATH/strings_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/decimals_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/executable_dictionary.xml $DEST_SERVER_PATH/
 
+# custom TLD
+ln -sf $SRC_PATH/top_level_domains /var/lib/clickhouse/
```
Comment (akuzm) [inline tests/config/install.sh:46]: Should probably be `$DEST_SERVER_PATH` here. AFAIK fasttest uses `install.sh` with proper arguments, so it will work as well.

Comment (azat) [reply tests/config/install.sh]: Right and require path replace, already done

**Comment by azat** [comment]
>Performance — Failed to parse the report.

Requires #17907

File: tests/performance/url_hits.xml [hunk 7]
```diff
@@ -32,6 +31,12 @@
            </values>
        </substitution>
     </substitutions>
-
     <query>SELECT count() FROM hits_100m_single WHERE NOT ignore({func}(URL))</query>
+
+    <!-- firstSignificantSubdomain/firstSignificantSubdomainCustom -->
```
Comment (akuzm) [inline tests/performance/url_hits.xml:36]: (cutTo)firstSignificantSubdomain(URL) is already tested by the above query, so I think we can leave just the Custom versions here, and use the same hits_100m_single table, for easier comparison.

File: docker/test/performance-comparison/config/config.d/top_level_domains_path.xml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<yandex>
+    <top_level_domains_path replace="replace">/top_level_domains/</top_level_domains_path>
```
Comment (akuzm) [inline docker/test/performance-comparison/config/config.d/top_level_domains_path.xml:2]: Why another file and not just add it to tweaks? This looks like an absolute path which does not exist, is it going to work? The configs end up at `{right,left}/db/config/top_level_domains`, not sure how we can specify this path...

Comment (akuzm) [reply docker/test/performance-comparison/config/config.d/top_level_domains_path.xml:2]: https://github.com/ClickHouse/ClickHouse/pull/11711

There is something weird going on with these paths... maybe we'll have to add another command line argument.

File: tests/performance/url_hits.xml (hunk 7 above)
Comment (azat) [reply tests/performance/url_hits.xml:36]: I don't see any point in comparing those functions with multiple threads, hence they were duplicated with max_threads=1 and smaller hits table.

And I decided to leave original queries as-is for now to avoid changes in perf tests

File: docker/test/performance-comparison/config/config.d/top_level_domains_path.xml (hunk 8 above)
Comment (azat) [reply docker/test/performance-comparison/config/config.d/top_level_domains_path.xml:2]: >Why another file and not just add it to tweaks?

For no reason

> This looks like an absolute path which does not exist, is it going to work? The configs end up at {right,left}/db/config/top_level_domains, not sure how we can specify this path...

It does, since it copies everything inside the container from the `docker/test/performance-comparison/config/`

https://github.com/ClickHouse/ClickHouse/blob/f6f336ac7839ed52a147eb0bcae8b88f0548082a/docker/test/performance-comparison/Dockerfile#L40

But it does not dereference symlinks, and docker will not do this, since it cannot leave the context, will address this

>There is something weird going on with these paths... maybe we'll have to add another command line argument.

Interesting. But this overrides will work, since it is absolute path, but ok, let's move it to the command line arguments.

**Comment by azat** [comment]
Can someone put force-tests label please?
(test failure reported to #17895)

**Comment by azat** [comment]
@akuzm looks like they haven't continued? some CI issues?

**Comment by azat** [comment]
>It's weird that we don't have the Partial Queries table with queries with new functions. I'll look into it later.

@akuzm Finally performance tests passed and has new functions in Partial Queries table.

**Review by akuzm — APPROVED** [review]
